WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,765 @@
+     * like "commons.wikimedia.org" and "wikidata.org" for multimedia/metadata-related in-app analytics.
+     *
+     * Cases where instrumentation would set a `domain`:
+     * - reading or editing an article
+     * - managing watchlist
+     * - interacting with feed
+     * - searching
+     *
+     * Cases where it might not be necessary for instrumentation to set a `domain`:
+     * - changing settings
+     * - managing reading lists
+     * - navigating map of nearby articles
+     * - multi-lingual features like Suggested Edits
+     * - marking session start/end; in which case schema and `data` should have a `languages` field
+     *   where user's list of languages can be stored, although it might make sense to set it to the domain
+     *   associated with the user's 1st preferred language – in which case use
+     *   `MWKLanguageLinkController.sharedInstance().appLanguage?.siteURL()!.host!`
+     */
+    @objc public func log(stream: String, schema: String, data: [String: NSCoding], domain: String? = nil) -> Void {
+        guard loggingEnabled, storageManager.sharingUsageData else {
+            return
+        }
+
+        let metaKey = "meta"
+        var meta: [String: NSCoding] = data[metaKey] as? [String: NSCoding] ?? [:]
+
+        if let domain = domain {
+            meta["domain"] = domain as NSCoding
+        }
+
+        var data = data
+        data[metaKey] = meta as NSCoding
+
+        /*
+         * The top-level field `client_dt` is for recording the time the event
+         * was generated. EventGate sets `meta.dt` during ingestion, so for
+         * analytics events that field is used as "timestamp of reception" and
+         * is used for partitioning the events in the database. See Phab:T240460
+         * for more information.
+         */
+        let clientDtKey = "client_dt"
+        if !data.keys.contains(clientDtKey) {
+            let clientDateTime = Date()
+            data[clientDtKey] = iso8601Formatter.string(from: clientDateTime) as NSCoding
+            lastTimestamp = clientDateTime
+        }
+
+        /*
+         * Generated events have the session ID attached to them before stream
+         * config is available (in case they're generated offline) and before
+         * they're cc'd to any other streams (once config is available).
+         */
+        let sessionIDKey = "session_id"
+        if !data.keys.contains(sessionIDKey) {
+            data[sessionIDKey] = sessionID as NSCoding
+        }
+
+        #if DEBUG
+        do {
+            let jsonString = try data.toPrettyPrintJSONString()
+            DDLogDebug("[EPC] Event logged to stream '\(stream)' (schema: \(schema):\n\(jsonString)")
+        } catch let error {
+            DDLogError("[EPC] \(error.localizedDescription)")
+        }
+        #endif
+
+        guard let streamConfigs = streamConfigurations else {
+            let event = EPCBufferEvent(stream: stream, schema: schema, data: data, domain: domain)
```

Comment:
@bearloga do you think we should persist a timestamp within EPCBufferEvent? Otherwise when it eventually does get posted because streamConfigurations is populated, I think the `client_dt` field will represent the post date but not the date the event was triggered.

Reply:
We should be good as is! When a buffered event gets re-`log()`-ed once stream configurations are available, that event's `data` will already have have a `client_dt` (and a `session_id`) populated from when the event was originally logged, before it was put into the input buffer.

Reply:
Without looking too closely, I think @tonisevener may be alluding to the issue where the timestamp is not generated until the event is being processed (i.e., pulled out of the input buffer). The time at which the event actually *happens* in software will thus be offset by the time it spends in the buffer. IMO the best option here is to just leave it simple, as it's a identified source of timestamp bias that we're going to correct.

Reply:
Oh, I see, yes here it's already been set in `data`. All good then.

Reply:
@bearloga @linehan 👍  doh, I didn't catch that part. I see it now, thanks for explaining!